src/main/java/org/openrewrite/staticanalysis/ReplaceWeekYearWithYear.java

```diff
@@ -0,0 +1,120 @@
+ * limitations under the License.
+ */
+package org.openrewrite.staticanalysis;
+
+import org.openrewrite.*;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.MethodMatcher;
+import org.openrewrite.java.search.UsesType;
+import org.openrewrite.java.tree.*;
+
+import java.util.*;
+
+public class ReplaceWeekYearWithYear extends Recipe {
+    public static final MethodMatcher SIMPLE_DATE_FORMAT_CONSTRUCTOR_MATCHER = new MethodMatcher("java.text.SimpleDateFormat <constructor>(..)");
+    public static final MethodMatcher OF_PATTERN_MATCHER = new MethodMatcher("java.time.format.DateTimeFormatter ofPattern(..)");
+
+    @Override
+    public String getDisplayName() {
+        return "Week Year (YYYY) should not be used for date formatting";
+    }
+
+    @Override
+    public String getDescription() {
+        return "For most dates Week Year (YYYY) and Year (yyyy) yield the same results. However, on the last week of" +
+               " December and first week of January Week Year could produce unexpected results.";
+    }
+
+    @Override
+    public Set<String> getTags() {
+        return Collections.singleton("RSPEC-3986");
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return Preconditions.check(
+                Preconditions.or(
+                        new UsesType<>("java.util.Date", false),
+                        new UsesType<>("java.time.format.DateTimeFormatter", false),
+                        new UsesType<>("java.text.SimpleDateFormat", false)
+                ),
+                new ReplaceWeekYearVisitor()
+        );
+    }
+
+    private static class ReplaceWeekYearVisitor extends JavaIsoVisitor<ExecutionContext> {
+        @Override
+        public J.MethodInvocation visitMethodInvocation(J.MethodInvocation mi, ExecutionContext ctx) {
+            if (OF_PATTERN_MATCHER.matches(mi)) {
+                getCursor().putMessage("KEY", mi);
+            }
+
+            return super.visitMethodInvocation(mi, ctx);
+        }
+
+        @Override
+        public J.NewClass visitNewClass(J.NewClass nc, ExecutionContext ctx) {
+            if (SIMPLE_DATE_FORMAT_CONSTRUCTOR_MATCHER.matches(nc)) {
+                getCursor().putMessage("KEY", nc);
+            }
+
+            return super.visitNewClass(nc, ctx);
+        }
+
+        @Override
+        public J.Literal visitLiteral(J.Literal li, ExecutionContext ctx) {
+            if (li.getValue() instanceof String) {
+                Cursor c = getCursor().dropParentWhile(is -> is instanceof J.Parentheses || !(is instanceof Tree));
+                if (c.getMessage("KEY") != null) {
+                    String value = li.getValueSource();
+
+                    if (value == null) {
+                        return li;
+                    }
+
+                    String newValue = replaceY(value);
+
+                    return li.withValueSource(newValue).withValue(newValue);
```

Comment:
This looks a bit wrong to me. The `valueSource` will be the value with quotes (possibly even Java 17 block quotes, for which we should have a test probably), while the `value` is without quotes. We should update it correctly here.